sdks/uniswapx-sdk/src/trade/V3DutchOrderTrade.ts

```diff
@@ -0,0 +1,159 @@
+        if (this._inputAmount) return this._inputAmount
+
+        const amount = CurrencyAmount.fromRawAmount(
+            this._currencyIn,
+            this.order.info.input.startAmount.toString()
+        )
+        this._inputAmount = amount
+        return amount
+    }
+
+    public get outputAmounts(): CurrencyAmount<TOutput>[] {
+        if (this._outputAmounts) return this._outputAmounts
+
+        const amounts = this.order.info.outputs.map((output) => {
+            // assume single chain ids across all outputs for now
+            const currencyOut = this._currenciesOut.find((currency) =>
+                areCurrenciesEqual(currency, output.token, currency.chainId)
+            )
+
+            if (!currencyOut) {
+                throw new Error("Currency out not found")
+            }
+
+            return CurrencyAmount.fromRawAmount(currencyOut, output.startAmount.toString())
+        })
+
+        this._outputAmounts = amounts
+        return amounts
+    }
+
+    private _firstNonFeeOutputStartEndAmounts:
+        | { startAmount: CurrencyAmount<TOutput>; endAmount: CurrencyAmount<TOutput> }
+        | undefined;
+
+    private getFirstNonFeeOutputStartEndAmounts(): {
+        startAmount: CurrencyAmount<TOutput>
+        endAmount: CurrencyAmount<TOutput>
+    } {
+        if (this._firstNonFeeOutputStartEndAmounts)
+            return this._firstNonFeeOutputStartEndAmounts;
+      
+        if (this.order.info.outputs.length === 0) {
+            throw new Error("there must be at least one output token");
+        }
+        const output = this.order.info.outputs[0];
+    
+        // assume single chain ids across all outputs for now
+        const currencyOut = this._currenciesOut.find((currency) =>
+            areCurrenciesEqual(currency, output.token, currency.chainId)
+        );
+    
+        if (!currencyOut) {
+        throw new Error(
+            "currency output from order must exist in currenciesOut list"
+        );
+        }
+
+        const startAmount = CurrencyAmount.fromRawAmount(currencyOut, output.startAmount.toString())
+        const nonFeeOutputEndAmount = getEndAmount({
+            startAmount: output.startAmount,
+            relativeAmounts: output.curve.relativeAmounts,
+        });
+        const endAmount = CurrencyAmount.fromRawAmount(currencyOut, nonFeeOutputEndAmount.toString())
+
+        this._firstNonFeeOutputStartEndAmounts = { startAmount, endAmount }
+        return { startAmount, endAmount }
+    }
+    
+     // TODO: revise when there are actually multiple output amounts. for now, assume only one non-fee output at a time
+  public get outputAmount(): CurrencyAmount<TOutput> {
+    return this.getFirstNonFeeOutputStartEndAmounts().startAmount;
+  }
+
+  public minimumAmountOut(): CurrencyAmount<TOutput> {
+    return this.getFirstNonFeeOutputStartEndAmounts().endAmount;
+  }
+
+  public maximumAmountIn(): CurrencyAmount<TInput> {
+    const endAmount = getEndAmount({
```

Comment:
We shouldn't assume that the last value is the max input.

Reply:
Same for `minimumAmountOut()`

Reply:
Rewrote these implementations with extra tests [here](https://github.com/Uniswap/sdks/pull/84/commits/438218183e3a4f47c428c0752f6d4dbe19592d83)!